src/catalog/src/durable/impls/persist.rs

```diff
@@ -786,6 +863,42 @@ impl PersistCatalogState {
         RelationDesc::empty().with_column("data", ScalarType::Jsonb.nullable(false))
     }
 
+    /// Increment the version in the upgrade shard to the code's current version.
+    async fn increment_catalog_upgrade_shard_version(&mut self, organization_id: Uuid) {
+        let upgrade_shard_id = shard_id(organization_id, UPGRADE_SEED);
+        let mut write_handle: WriteHandle<(), (), Timestamp, Diff> = self
+            .persist_client
+            .open_writer(
+                upgrade_shard_id,
+                Arc::new(UnitSchema::default()),
+                Arc::new(UnitSchema::default()),
+                Diagnostics {
+                    shard_name: UPGRADE_SHARD_NAME.to_string(),
+                    handle_purpose: "increment durable catalog upgrade shard version".to_string(),
+                },
+            )
+            .await
+            .expect("invalid usage");
+        const EMPTY_UPDATES: &[(((), ()), Timestamp, Diff)] = &[];
+        let mut upper = write_handle.fetch_recent_upper().await.clone();
+        loop {
+            let next_upper = upper
+                .iter()
+                .map(|timestamp| timestamp.step_forward())
+                .collect();
+            match write_handle
+                .compare_and_append(EMPTY_UPDATES, upper, next_upper)
+                .await
+                .expect("invalid usage")
+            {
+                Ok(()) => break,
+                Err(upper_mismatch) => {
+                    upper = upper_mismatch.current;
```

Comment:
We don't seem to have a test for this error case.

Reply:
This involves a very tricky to repro race condition and I don't think I can write a test that is guaranteed to hit this or is even likely to hit this.